controllers/util.go

```diff
@@ -17,32 +19,122 @@ import (
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
 )
 
 type StateStore interface {
 	client.Object
 	GetSecretRef() *v1.SecretReference
 }
 
+type opts struct {
+	ctx    context.Context
+	client client.Client
+	logger logr.Logger
+}
+
+type jobOpts struct {
+	opts
+	obj               *unstructured.Unstructured
+	pipelineLabels    map[string]string
+	pipelineResources []client.Object
+}
+
+func ensurePipelineIsReconciled(j jobOpts) (*ctrl.Result, error) {
+	namespace := j.obj.GetNamespace()
+	if namespace == "" {
+		namespace = kratixPlatformSystemNamespace
+	}
+
+	pipelineJobs, err := getJobsWithLabels(j.opts, j.pipelineLabels, namespace)
+	if err != nil {
+		j.logger.Info("Failed getting Promise pipeline jobs", "error", err)
+		return &slowRequeue, nil
+	}
+
+	// No jobs indicates this is the first reconciliation loop of this resource request
+	if len(pipelineJobs) == 0 {
+		j.logger.Info("No jobs found, creating workflow Job")
+		return &fastRequeue, createConfigurePipeline(j)
+	}
+
+	existingPipelineJob, err := resourceutil.PipelineExists(j.logger, j.obj, pipelineJobs)
+	if err != nil {
+		return &slowRequeue, nil
+	}
+
+	if resourceutil.IsThereAPipelineRunning(j.logger, pipelineJobs) {
+		for _, job := range resourceutil.SuspendablePipelines(j.logger, pipelineJobs) {
```

Comment:
🤔: is this a check we only wanna do if there's a pipeline running? wouldn't we want to suspend all pipelines that are not for the most recent spec?

Reply:
We only *need* to suspend if one is running, if none are inflight (all jobs are complete/suspened/failed) its okay not to.